src/util/sighash.rs

```diff
@@ -0,0 +1,909 @@
+    SinglePlusAnyoneCanPay = 0x83,
+
+    /// Reserved for future use, `#[non_exhaustive]` is not available with current MSRV
+    Reserved = 0xFF,
+}
+
+/// Possible errors in computing the signature message
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum Error {
+    /// Could happen only by using `*_encode_signing_*` methods with custom writers, engines writers
+    /// like the ones used in methods `*_signature_hash` don't error
+    Io(io::ErrorKind),
+
+    /// Requested index is greater or equal than the number of inputs in the transaction
+    IndexOutOfInputsBounds {
+        /// Requested index
+        index: usize,
+        /// Number of transaction inputs
+        inputs_size: usize,
+    },
+
+    /// Using SIGHASH_SINGLE without a "corresponding output" (an output with the same index as the
+    /// input being verified) is a validation failure
+    SingleWithoutCorrespondingOutput {
+        /// Requested index
+        index: usize,
+        /// Number of transaction outputs
+        outputs_size: usize,
+    },
+
+    /// There are mismatches in the number of prevouts provided compared with the number of
+    /// inputs in the transaction
+    PrevoutsSize,
+
+    /// Requested a prevout index which is greater than the number of prevouts provided or a
+    /// [`Prevouts::One`] with different index
+    PrevoutIndex,
+
+    /// A single prevout has been provided but all prevouts are needed without `ANYONECANPAY`
+    PrevoutKind,
+
+    /// Annex must be at least one byte long and the first bytes must be `0x50`
+    WrongAnnex,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::Io(ref e) => write!(f, "Writer errored: {:?}", e),
+            Error::IndexOutOfInputsBounds { index, inputs_size } => write!(f, "Requested index ({}) is greater or equal than the number of transaction inputs ({})", index, inputs_size),
+            Error::SingleWithoutCorrespondingOutput { index, outputs_size } => write!(f, "SIGHASH_SINGLE for input ({}) haven't a corresponding output (#outputs:{})", index, outputs_size),
+            Error::PrevoutsSize => write!(f, "Number of supplied prevouts differs from the number of inputs in transaction"),
+            Error::PrevoutIndex => write!(f, "The index requested is greater than available prevouts or different from the provided [Provided::Anyone] index"),
+            Error::PrevoutKind => write!(f, "A single prevout has been provided but all prevouts are needed without `ANYONECANPAY`"),
+            Error::WrongAnnex => write!(f, "Annex must be at least one byte long and the first bytes must be `0x50`"),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl ::std::error::Error for Error {}
+
+impl<'u> Prevouts<'u> {
+    fn check_all(&self, tx: &Transaction) -> Result<(), Error> {
+        if let Prevouts::All(prevouts) = self {
```

Comment:
nit: as noted earlier, a non-slice type would eliminate this check.

Reply:
This isn't a check, it's a destructuring.

Reply:
Ah the check is the line below